extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java

```
@@ -1016,22 +1091,66 @@ static class StringTransfer implements TransferObject<Binary[]> {
+
+        /**
+         * Total number of chunks buffered so far
+         */
+        private int numChunksBuffered;
 
-        StringTransfer(ColumnSource<?> columnSource, int targetSize) {
+
+        StringTransfer(ColumnSource<?> columnSource, int maxValuesPerPage, int targetPageSize) {
             this.columnSource = columnSource;
-            this.buffer = new Binary[targetSize];
-            context = this.columnSource.makeGetContext(targetSize);
+            this.buffer = new Binary[maxValuesPerPage];
+            context = this.columnSource.makeGetContext(maxValuesPerPage);
+            numChunksBuffered = 0;
+            this.maxValuesPerPage = maxValuesPerPage;
+            this.targetPageSize = targetPageSize;
         }
 
         @Override
-        public void propagateChunkData() {
-            for (int i = 0; i < chunk.size(); i++) {
-                String value = chunk.get(i);
-                buffer[i] = value == null ? null : Binary.fromString(value);
+        public boolean hasMoreDataToBuffer() {
+            return ((chunk != null) && (numChunksBuffered < chunk.size()));
+        }
+
+        public int bufferFetchedData() {
+            if (numChunksBuffered != 0) {
+                // Clear any old data in the buffer from previous iterations
+                buffer = new Binary[maxValuesPerPage];
             }
+            int bufferedDataSize = 0;
+            int bufferIdx = 0; // Stores the number of chunks buffered in this iteration
+            while (numChunksBuffered < chunk.size()) {
+                final String value = chunk.get(numChunksBuffered);
+                if (value == null) {
+                    buffer[bufferIdx++] = null;
+                } else {
+                    Binary binaryVal = Binary.fromString(value);
+                    if (binaryVal.length() > targetPageSize) {
+                        throw new UncheckedDeephavenException(String.format("Cannot fit string data of size %d in a " +
```

Do we need to throw an exception or can we write the page and continue? Also, consider the case where we have a single string >= targetPageSize, shouldn't we write an abnormal page size and continue writing?